concordium-std-derive/src/lib.rs

```diff
@@ -1202,6 +1224,154 @@ fn schema_type_fields(fields: &syn::Fields) -> syn::Result<proc_macro2::TokenStr
     }
 }
 
+/// We reserve a number of error codes for custom errors, such as ParseError,
+/// that are provided by concordium-std. The error codes for user-defined error
+/// types will have this number as an offset.
+const RESERVED_ERROR_CODES: u32 = 100;
+
+/// Derive the conversion of enums that represent error types into the Reject
+/// struct which can be used as the error type of init and receive functions.
+/// Creating custom enums for error types can provide meaningful error messages
+/// to the user of the smart contract.
+///
+/// Please note:
+///   - At the moment, we can only derive fieldless enums.
+///   - The error-type enum needs to derive (or implement) the Copy and Clone
+///     traits.
+///
+/// ### Example
+/// ```ignore
+/// #[derive(Clone, Copy, Reject)]
+/// enum MyError {
+///     IllegalState,
+///     WrongSender,
+///     // TimeExpired(time: Timestamp), /* currently not supported */
+///     ...
+/// }
+/// ```ignore
+/// #[receive(contract = "my_contract", name = "some_receive")]
+/// fn receive<A: HasActions>(ctx: &impl HasReceiveContext, state: &mut MyState)
+/// -> Result<A, MyError> {...} ```
+#[proc_macro_derive(Reject, attributes(from))]
+pub fn reject_derive(input: TokenStream) -> TokenStream {
+    unwrap_or_report(reject_derive_worker(input))
+}
+
+fn reject_derive_worker(input: TokenStream) -> syn::Result<TokenStream> {
+    let ast: syn::DeriveInput = syn::parse(input)?;
+    let enum_data = match ast.data {
+        syn::Data::Enum(data) => Ok(data),
+        _ => Err(syn::Error::new(ast.span(), "Reject can only be derived for enums.")),
+    }?;
+    let enum_ident = ast.ident;
+
+    let variant_error_conversions: Vec<_> =
+        generate_variant_error_conversions(&enum_data, &enum_ident);
+
+    let exceeded_error_code_limit = format!(
+        "Error code should not exceed {} (i32::MAX minus reserved error-code offset).",
+        i32::MAX - RESERVED_ERROR_CODES as i32
+    );
+
+    let gen = quote! {
+        impl From<#enum_ident> for Reject {
+            fn from(e: #enum_ident) -> Self {
+                Reject { error_code: (e as u32 + 1).checked_add(#RESERVED_ERROR_CODES).expect(#exceeded_error_code_limit) }
```

Comment:
If we make the reserved codes as I suggested then we can avoid this wart here with adding an offset.

Reply:
Maybe it would be worthwhile to also add an `#[inline]` annotation on the generated function.

We should test though what, if any, effect it has on the generated size of the `.wasm` file.

Reply:
For the `error_code` smart contract, it didn't have any effect. Both produced 1,104 bytes.